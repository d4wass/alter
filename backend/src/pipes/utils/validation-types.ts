import { ValidationError } from '@nestjs/common';

export type ConvertedValidationError = {
  property: string;
  constraints: any;
  children?: ValidationError[];
};

export type ConstructorTypes =
  | StringConstructor
  | BooleanConstructor
  | NumberConstructor
  | ArrayConstructor
  | ObjectConstructor;
