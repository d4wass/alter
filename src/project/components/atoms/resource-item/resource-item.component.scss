@use '../../../../base';
@use '../../../../mixins';

@include mixins.paragraph;

.wrapper,
.content{
    display: flex;
    flex-direction: column;
    align-items: center;
}

.wrapper {
    justify-content: space-between;
    max-width: 300px;
    height: 400px;
    padding: 30px 20px;
    margin: 20px;
    border-radius: base.$main-border-rad;
    box-shadow: 1px 3px 7px rgba(0, 0, 0, 0.25);
    background-color: #fff;
}

.content{
    height: 100%;
    h2, h5{
        margin:0;
    }

    h5{
        margin: 10px 0 0 0;
    }

    p{
        margin: 0;
        margin-top: 20px;
    }

}

a{
    margin: 20px 0;
    cursor: pointer;
    transition: all ease-in-out .3s;
    &:hover{
        color: base.$yellow-span
    }
}

img{
    width: auto;
    height: 30px;
    margin-bottom: 30px;
}