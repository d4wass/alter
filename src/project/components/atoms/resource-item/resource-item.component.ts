import { Component, Input } from '@angular/core';

export interface CtaBtn {
  content: string;
  route: string;
}

@Component({
    selector: 'app-resource-item',
    template: `
    <div class="wrapper">
      <img src="{{ icon }}" />
      <div class="content">
        <h2>{{ title }}</h2>
        <h5 *ngIf="subtitle" data-test="subtitle">{{ subtitle }}</h5>
        <p>{{ content }}</p>
      </div>
      <a routerLink="#">{{ cta?.content }}</a>
    </div>
  `,
    styleUrls: ['./resource-item.component.scss'],
    standalone: false
})
export class ResourceItemComponent {
  @Input() title!: string;
  @Input() subtitle?: string;
  @Input() content?: string;
  @Input() cta?: CtaBtn;
  @Input() icon!: string;
}
