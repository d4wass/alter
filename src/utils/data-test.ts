export function dataTest(selector: string): string {
  return selector && `[data-test="${selector}"]`;
}
