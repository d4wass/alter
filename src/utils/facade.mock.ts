import { Observable, BehaviorSubject } from 'rxjs';

export type FacadeMock<T> = {
  [P in keyof T]: T[P] extends Observable<infer U>
    ? BehaviorSubject<U>
    : T[P] extends (args: infer A) => Observable<infer R>
    ? (args: A) => BehaviorSubject<R>
    : T[P];
};
