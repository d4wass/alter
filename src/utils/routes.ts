export interface urlItem {
  description: string;
  url: string;
}

export const urlList: urlItem[] = [
  { description: 'Learn more', url: 'faq' },
  { description: 'Carculator', url: 'carculator' },
  { description: 'Become a host', url: 'host' },
  { description: 'Create Account', url: 'new-user' },
  { description: 'Your Cars', url: 'vehicles' }
];
