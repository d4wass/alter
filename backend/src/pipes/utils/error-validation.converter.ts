import { ValidationError } from 'class-validator';
import { ConvertedValidationError } from './validation-types';

export class ValidationErrorConverter {
  private _convertedErrors: Array<ConvertedValidationError> = [];
  public get convertedErrors(): Array<ConvertedValidationError> {
    return this._convertedErrors;
  }
  public set convertedErrors(value: Array<ConvertedValidationError>) {
    this._convertedErrors = value;
  }
  private _errors: ValidationError[];

  constructor(errors: ValidationError[]) {
    this._errors = errors;
    this.errorsConverter(this._errors);
  }

  errorsConverter(errors: ValidationError[]): void {
    errors.forEach((error) => {
      if (!error.children.length) {
        this._convertedErrors.push({ property: error.property, constraints: error.constraints });
      } else {
        this.errorsConverter(error.children);
      }
    });
  }
}
