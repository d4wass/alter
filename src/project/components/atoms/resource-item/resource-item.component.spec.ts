import { createComponentFactory, Spectator } from '@ngneat/spectator';
import { dataTest } from '../../../../utils/data-test';
import { ObjectType } from 'typescript';
import { CtaBtn, ResourceItemComponent } from './resource-item.component';

describe('ResourceItemComponent', () => {
  let component: ResourceItemComponent;
  let spectator: Spectator<ResourceItemComponent>;

  const createComponent = createComponentFactory({
    component: ResourceItemComponent
  });

  const selectors = {
    subtitle: () => spectator.query(dataTest('subtitle'))
  };

  beforeEach(() => {
    spectator = createComponent();
    component = spectator.component;
  });

  it('should match to snapshot', () => {
    expect(spectator.fixture).toMatchSnapshot();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  describe('@Inputs', () => {
    it('should set value of (content)', () => {
      component.content = 'value';
      expect(component.content).toEqual('value');
    });
    it('should set value of (title)', () => {
      component.title = 'value';
      expect(component.title).toEqual('value');
    });
    it('should set value of (subtitle)', () => {
      component.subtitle = 'value';
      expect(component.subtitle).toEqual('value');
    });
    it('should set value of (icon)', () => {
      component.icon = 'value';
      expect(component.icon).toEqual('value');
    });
    it('should set value of (ctaBtn)', () => {
      component.cta = { content: 'value', route: 'route' };
      expect(component.cta).toEqual({ content: 'value', route: 'route' });
    });
  });
  describe('Template', () => {
    it('should display subtitle if have value', () => {
      component.subtitle = 'subtitle';
      spectator.detectChanges();

      expect(selectors.subtitle()?.innerHTML).toEqual('subtitle');
    });
    it('should not display subtitle if have not value', () => {
      expect(selectors.subtitle()?.innerHTML).toBeUndefined();
    });
  });
});
