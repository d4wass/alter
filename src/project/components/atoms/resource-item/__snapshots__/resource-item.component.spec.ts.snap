// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResourceItemComponent should match to snapshot 1`] = `
<app-resource-item>
  <div
    class="wrapper"
  >
    <img
      src=""
    />
    <div
      class="content"
    >
      <h2>
      </h2>
      <p>
      </p>
    </div>
    <a
      routerlink="#"
    >
    </a>
  </div>
</app-resource-item>
`;
